File: src/python/nispor/route_rule.py
# SPDX-License-Identifier: Apache-2.0


class NisporRouteRuleState:
    def __init__(self, info):
        self._rls = [NisporRouteRule(rl_info) for rl_info in info]

    def __iter__(self):
        for rl in self._rls:
            yield rl


class NisporRouteRule:
    def __init__(self, info):
        self._info = info

    @property
    def action(self):
        return self._info["action"]

    @property
    def address_family(self):
        return self._info["address_family"]

    @property
    def flags(self):
        return self._info["flags"]

    @property
    def tos(self):
        return self._info["tos"]

    @property
    def table(self):
        return self._info.get("table")

    @property
    def dst(self):
        return self._info.get("dst")

    @property
    def src(self):
        return self._info.get("src")

    @property
    def iif(self):
        return self._info.get("iif")

    @property
    def oif(self):
        return self._info.get("oif")

    @property
    def goto(self):
        return self._info.get("goto")

    @property
    def priority(self):
        return self._info.get("priority")

    @property
    def fw_mark(self):
        return self._info.get("fw_mark")

    @property
    def fw_mask(self):
        return self._info.get("fw_mask")

    @property
    def mask(self):
        return self._info.get("mask")

    @property
    def flow(self):
        return self._info.get("flow")

    @property
    def tun_id(self):
        return self._info.get("tun_id")

    @property
    def suppress_ifgroup(self):
        return self._info.get("suppress_ifgroup")

    @property
    def suppress_prefix_len(self):
        return self._info.get("suppress_prefix_len")

    @property
    def protocol(self):
        return self._info.get("protocol")

    @property
    def ip_proto(self):
        return self._info.get("ip_proto")

    @property
    def src_port_range(self):
        return self._info.get("src_port_range")

    @property
    def dst_port_range(self):
        return self._info.get("dst_port_range")

    @property
    def l3mdev(self):
        return self._info.get("l3mdev", False)

File: src/lib/integ_tests/route_rule.rs
// SPDX-License-Identifier: Apache-2.0

use crate::NetState;

use std::panic;

use super::utils::assert_value_match;

const TEST_TABLE_ID: u32 = 100;

const EXPECTED_YAML_OUTPUT: &str = r#"---
- action: blackhole
  address_family: ipv6
  flags: 0
  tos: 0
  table: 100
  dst: 2001:db8:f::252/128
  src: 2001:db8:f::255/128
  iif: eth1
  oif: eth2
  priority: 998
- action: table
  address_family: ipv6
  flags: 0
  tos: 16
  table: 100
  dst: "2001:db8:f::253/128"
  src: "2001:db8:f::254/128"
  iif: eth1
  oif: eth2
  priority: 999
- action: unreachable
  address_family: ipv4
  flags: 0
  tos: 0
  dst: 192.0.2.2/32
  src: 192.0.2.1/32
  iif: eth1
  oif: eth2
  priority: 998
- action: table
  address_family: ipv4
  flags: 0
  tos: 16
  table: 100
  dst: 192.0.2.2/32
  src: 192.0.2.1/32
  iif: eth1
  oif: eth2
  priority: 999"#;

#[test]
fn test_get_route_rule_yaml() {
    with_route_rule_test_iface(|| {
        let state = NetState::retrieve().unwrap();
        let mut expected_rules = Vec::new();
        for mut rule in state.rules {
            if Some(TEST_TABLE_ID) == rule.table {
                // Travis CI Ubuntu 18.04 does not support protocol.
                rule.protocol = None;
                expected_rules.push(rule)
            }
        }
        assert_value_match(EXPECTED_YAML_OUTPUT, &expected_rules);
    });
}

fn with_route_rule_test_iface<T>(test: T)
where
    T: FnOnce() + panic::UnwindSafe,
{
    super::utils::set_network_environment("rule");

    let result = panic::catch_unwind(|| {
        test();
    });

    super::utils::clear_network_environment();
    assert!(result.is_ok())
}
